### PR TITLE
Implement keep_history flag

### PR DESCRIPTION
**Pull request by ghost**
See https://github.com/travis-ci/travis-ci/issues/7562

I find force pushing and rewriting history to be unintuitive behavior. I noticed that there is a `keep_history` flag in the spec and wanted to try a hand at implementing it. I have never programmed in Ruby before so apologies if there are problems.

### CODE REVIEW

File: lib/dpl/provider/pages.rb [hunk 1]
```diff
@@ -49,15 +51,22 @@ def needs_key?
...
-        context.shell 'git init' or raise 'Could not create new git repo'
         context.shell "git config user.email '#{@gh_email}'"
         context.shell "git config user.name '#{@gh_name}'"
         context.shell "echo '#{@gh_fqdn}' > CNAME" if @gh_fqdn
-        context.shell 'git add .'
-        context.shell "git commit -m 'Deploy #{@project_name} to #{@gh_ref}:#{@target_branch}'"
-        context.shell "git push --force --quiet 'https://#{@gh_token}@#{@gh_ref}' master:#{@target_branch} > /dev/null 2>&1"
+
+        if @keep_history
+          context.shell 'git add .'
```
Comment (webknjaz) [inline lib/dpl/provider/pages.rb:60]: This has to be done on the target branch.

**Comment by webknjaz** [comment]
Hey @remyabel,

Thank you for your effort!
I've explained the overall situation in travis-ci/travis-ci#7562. I don't know Ruby either :)

The resulting implementations should be way more complex. Please looks at the [`feature/enable-gh-pages-keep-history` branch](https://github.com/webknjaz/dpl/tree/feature/enable-gh-pages-keep-history) to learn my findings. It seems I've covered more cases there.

I'm testing it via [this repo](https://github.com/webknjaz/dpl-gh-pages), you may find example usage of experimental branch:
```yaml
...
deploy:
- edge:
    source: webknjaz/dpl
    branch: feature/enable-gh-pages-keep-history
  provider: pages
  keep-history: true
  verbose: true
  skip-cleanup: true
  github-token:
    secure: {{your_encrypted_key_here}}
  local-dir: build
  target-branch: multi-provider-gh-pages-nonexistent
  on:
    branch: master
...
```
**CAUTION:** There's issues https://travis-ci.org/webknjaz/dpl-gh-pages/builds/204541804#L204 with fs interaction — I think, the tmpdir becomes removed earlier then all git commands (including push) happen). *DO NOT TRY THIS IN PRODUCTION REPO.*

--
Warm regards,
Sviatoslav.

**Comment by LeafHacker** [comment]
Any update on this? Would be great to have this feature to avoid uploading large repos for small changes.

**Comment by webknjaz** [comment]
@LeafHacker I'd gladly accept help from some Ruby person, why would explain my why some weird race condition happens within this deployment provider. Currently, draft version might be ready, but I won't submit it knowing that there's a known bug, which is hard (for me) to track down.

**Comment by webknjaz** [comment]
It's finally been well-coded via travis-ci/dpl#719

**Pull request closed without merging**